mdtraj/formats/pdbx/pdbxfile.py

```diff
@@ -0,0 +1,727 @@
+            print("#", file=file)
+
+        # Identify bonds that should be listed in the file.
+
+        bonds = []
+        for atom1, atom2 in topology.bonds:
+            if (
+                atom1.residue.name not in PDBxFile._standardResidues
+                or atom2.residue.name not in PDBxFile._standardResidues
+            ):
+                bonds.append((atom1, atom2))
+            elif (
+                atom1.name == "SG"
+                and atom2.name == "SG"
+                and atom1.residue.name == "CYS"
+                and atom2.residue.name == "CYS"
+            ):
+                bonds.append((atom1, atom2))
+        if len(bonds) > 0:
+            # Write the bond information.
+
+            print("loop_", file=file)
+            print("_struct_conn.id", file=file)
+            print("_struct_conn.conn_type_id", file=file)
+            print("_struct_conn.ptnr1_label_asym_id", file=file)
+            print("_struct_conn.ptnr1_label_comp_id", file=file)
+            print("_struct_conn.ptnr1_label_seq_id", file=file)
+            print("_struct_conn.ptnr1_label_atom_id", file=file)
+            print("_struct_conn.ptnr2_label_asym_id", file=file)
+            print("_struct_conn.ptnr2_label_comp_id", file=file)
+            print("_struct_conn.ptnr2_label_seq_id", file=file)
+            print("_struct_conn.ptnr2_label_atom_id", file=file)
+            chainIds = {}
+            resIds = {}
+            if keepIds:
+                for chain in topology.chains:
+                    chainIds[chain] = chain.id
+                for res in topology.residues:
+                    resIds[res] = res.id
+            else:
+                for chainIndex, chain in enumerate(topology.chains):
+                    chainIds[chain] = chr(ord("A") + chainIndex % 26)
+                    for resIndex, res in enumerate(chain.residues):
+                        resIds[res] = resIndex + 1
+            for i, (atom1, atom2) in enumerate(bonds):
+                if atom1.residue.name == "CYS" and atom2.residue.name == "CYS":
+                    bondType = "disulf"
+                else:
+                    bondType = "covale"
+                line = "bond%d %s %s %-4s %5s %-4s %s %-4s %5s %-4s"
+                print(
+                    line
+                    % (
+                        i + 1,
+                        bondType,
+                        chainIds[atom1.residue.chain],
+                        atom1.residue.name,
+                        resIds[atom1.residue],
+                        atom1.name,
+                        chainIds[atom2.residue.chain],
+                        atom2.residue.name,
+                        resIds[atom2.residue],
+                        atom2.name,
+                    ),
+                    file=file,
+                )
+            print("#", file=file)
+
+        # Write the header for the atom coordinates.
+
+        print("loop_", file=file)
```

Comment:
I'm sure doing 20+ print statements in a row probably involves opening and closing the file 20+ times. Might as well open the file once in a context manager + write instead.

Reply:
Ah, You're right! I assumed that `file` was the same as the `Pdbxfile.__init__ ` `file` argument. Please disregard...